pkg/render/applicationlayer/applicationlayer.go

```diff
@@ -0,0 +1,373 @@
+)
+
+func ApplicationLayer(
+	pullSecrets []*corev1.Secret,
+	installation *operatorv1.InstallationSpec,
+	osType rmeta.OSType,
+	logsEnabled bool,
+	logIntervalSeconds *int64,
+	logRequestsPerInterval *int64,
+) render.Component {
+	return &component{
+		pullSecrets:            pullSecrets,
+		installation:           installation,
+		osType:                 osType,
+		logsEnabled:            logsEnabled,
+		logIntervalSeconds:     logIntervalSeconds,
+		logRequestsPerInterval: logRequestsPerInterval,
+	}
+}
+
+type component struct {
+	pullSecrets            []*corev1.Secret
+	installation           *operatorv1.InstallationSpec
+	osType                 rmeta.OSType
+	proxyImage             string
+	collectorImage         string
+	envoyConfigMap         *corev1.ConfigMap
+	logsEnabled            bool
+	logRequestsPerInterval *int64
+	logIntervalSeconds     *int64
+}
+
+func (c *component) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := c.installation.Registry
+	path := c.installation.ImagePath
+	prefix := c.installation.ImagePrefix
+
+	if c.osType != c.SupportedOSType() {
+		return fmt.Errorf("l7 log collection is supported only on %s", c.SupportedOSType())
+	}
+
+	var err error
+	var errMsgs []string
+
+	c.proxyImage, err = components.GetReference(components.ComponentEnvoyProxy, reg, path, prefix, is)
+
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+	c.collectorImage, err = components.GetReference(components.ComponentL7Collector, reg, path, prefix, is)
+
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+
+	if len(errMsgs) != 0 {
+		return fmt.Errorf(strings.Join(errMsgs, ","))
+	}
+
+	return nil
+}
+
+func (c *component) SupportedOSType() rmeta.OSType {
+	return rmeta.OSTypeLinux
+}
+
+func (c *component) Objects() ([]client.Object, []client.Object) {
+	var objs []client.Object
+	// TODO: ensure that namespace exists before proceeding
+
+	// If l7spec is provided render the required objects.
+	c.envoyConfigMap = c.envoyL7ConfigMap()
+	objs = append(objs, c.serviceAccount())
+	objs = append(objs, c.envoyConfigMap)
+	objs = append(objs, c.daemonset())
+
+	if c.installation.KubernetesProvider == operatorv1.ProviderDockerEE {
+		objs = append(objs, c.clusterAdminClusterRoleBinding())
+	}
+
+	// If we're running on openshift, we need to add in an SCC.
+	if c.installation.KubernetesProvider == operatorv1.ProviderOpenShift {
+		objs = append(objs, c.securityContextConstraints())
+	}
+
+	// Delete all the objects if logs are not enabled.
+	if !c.logsEnabled {
+		return nil, objs
+	}
+
+	return objs, nil
+}
+
+func (c *component) Ready() bool {
+	return true
+}
+
+// daemonset creates a daemonset for the L7 log collector component.
+func (c *component) daemonset() *appsv1.DaemonSet {
+	maxUnavailable := intstr.FromInt(1)
+
+	annots := map[string]string{}
+
+	if c.envoyConfigMap != nil {
+		annots[EnvoyConfigMapName] = rmeta.AnnotationHash(c.envoyConfigMap)
+	}
+
+	podTemplate := &corev1.PodTemplateSpec{
```

Comment:
Nit (and there's nothing _really,_ wrong with this), but I don't think you need to create a reference for this object just to de-reference it.